docs/proposals/20191030-machine-health-checking.md

```diff
@@ -0,0 +1,221 @@
+---
+title: Machine health checking a.k.a node auto repair
+authors:
+  - "@enxebre"
+  - "@bison"
+reviewers:
+  - "@janedoe"
+creation-date: yyyy-mm-dd
+last-updated: yyyy-mm-dd
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Title
+- Machine health checking a.k.a node auto repair
+
+## Table of Contents
+
+A table of contents is helpful for quickly jumping to sections of a proposal and for highlighting
+any additional information provided beyond the standard proposal template.
+[Tools for generating](https://github.com/ekalinin/github-markdown-toc) a table of contents from markdown are available.
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [User Stories [optional]](#user-stories-optional)
+      - [Story 1](#story-1)
+      - [Story 2](#story-2)
+    - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Design Details](#design-details)
+    - [Test Plan](#test-plan)
+    - [Graduation Criteria](#graduation-criteria)
+      - [Examples](#examples)
+        - [Alpha -> Beta Graduation](#alpha---beta-graduation)
+        - [Beta -> GA Graduation](#beta---ga-graduation)
+        - [Removing a deprecated flag](#removing-a-deprecated-flag)
+    - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+    - [Version Skew Strategy](#version-skew-strategy)
+  - [Implementation History](#implementation-history)
+  - [Drawbacks [optional]](#drawbacks-optional)
+  - [Alternatives [optional]](#alternatives-optional)
+  - [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+Enable opt in automated health checking and remediation of unhealthy nodes backed by machines.
+
+## Motivation
+
+- Reduce administrative overhead to run a cluster.
+
+- Ability to respond to failures of machines and keep the cluster nodes healthy.
+
+### Goals
+
+- Enable automated remediation for pools of machines/nodes
+
+- Allow to define different unhealthy criterias based on node conditions for different pools.
+
+- Providing administrator-defined short-circuiting of automated remediation when multiple nodes are unhealthy at the same time.
+
+### Non-Goals/Future Work
```

Comment:
This only covers worker nodes? Might be good to add control plane remediation as non-goal if that is the intent

Reply:
I think this should apply fine to the control plane proposal (assuming that one defines the remediation threshold as a small percentage or number).

Reply:
This covers any arbitrary group of machines/nodes as a user you target in your CRD definition. Whereas Jason statement above is true, notice that:
- This not meant to provide strong guarantees that application quorum for N members is maintained at any time. This does not know about the app layer, that can be handled in conjunction with PDB to block deletion on drain or other mechanisms.
- This prevents farther remediation from happening when a threshold of unhealthy machines is reached, but it does not prevent the machines from becoming unhealthy in the first place, involuntary disruptions will happen nevertheless.

I added a non-goal to clarify this point.